### ISSUE TITLE
`no-implicit-global-access`

### ISSUE THREAD

**Issue by SNDST00M** [eslint/eslint#14905]
**Please describe what the rule should do:**

The new rule will prevent users from accessing global variables without a `window` or `global` prefix.

**What new ECMAScript feature does this rule relate to?**

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// This generates a error
const nogui = new URLSearchParams(window.location.search).get("nogui");

// This is silent
const nogui = new window.URLSearchParams(window.location.search).get("nogui");
```

**Why should this rule be included in ESLint (instead of a plugin)?**

- Complement `no-implicit-globals` but for read operations instead of write operations.
- Sometimes there may be a variable in an upper local scope that shadows a global.

**Are you willing to submit a pull request to implement this rule?**

If I get time, I guess.

**Comment by ljharb** [comment]
`window` is itself implicit global access, looking up the `window` property on the global. That's how all globals work.

**Comment by SNDST00M** [comment]
What would you call the rule? I think it still offers merit.

**Comment by ljharb** [comment]
It's not about the name - it's that *every single global*, including the necessary ones like `Array`, `Promise`, etc - are accessed this way.

A program shouldn't have any global variables besides the built-in ones anyways - and `no-undef` already covers that, and has for a long time. What's wrong with `new URLSearchParams`, for example?

**Comment by SNDST00M** [comment]
It's implicit.

In scripts designed for isolated contexts (e.g. userscripts, preload scripts in Electron), the context changes and only some of these globals are present.

This shouldn't be a problem thanks to ESLint environments, but in an IIFE with 2 upper function contexts, it becomes an issue. Global access is also necessary with user libraries as well as author scripts.

**Comment by ljharb** [comment]
It’s not implicit because it’s provided by the language/runtime, any more than `function () {}` or `[]` is implicit.

IIFEs are obsolete since modules, so I’m not sure why that would ever exist in modern code anymore.

Global access is never necessary with user libraries, because everyone should be using a module system and likely a bundler.

**Comment by SNDST00M** [comment]
That's good, and this rule should be a plugin then (third party or otherwise). Unfortunately there's a 4% of modern browsers that don't support EMCAScript modules, so I will stick with the IIFE for a year even though it's an ugly hack.

**Comment by ljharb** [comment]
Even with ES Modules, you'll still want a bundler for the foreseeable future; the design of ES Modules is always going to be slower (due to waterfall dependency discovery). A build process remains required for modern web dev.
